struct.go

```diff
@@ -0,0 +1,218 @@
+package tradingview
+
+import (
+	"errors"
+	"maps"
+	"net/http"
+	"slices"
+	"sync"
+
+	"github.com/gorilla/websocket"
+)
+
+/*
+Handles data associated with an instance of the websocket
+*/
+type TV_API struct {
+	ws *websocket.Conn
+
+	ReadCh          chan map[string]interface{}
+	writeCh         chan map[string]interface{}
+	ErrorCh         chan error // receives errors that occurred in read/write threads
+	internalErrorCh chan error // internal handling of errors in read/write threads
+
+	symbolCounter   uint64
+	seriesCounter   uint64
+	seriesCreated   bool // use sync.Once in clean up
+	resolvedSymbols map[string]string
+	seriesMap       map[string]string
+
+	csToken string
+	qsToken string
+	qs      string
+	qssq    string
+
+	realtimeSymbols map[string]bool
+
+	halts await_response
+}
+
+/*
+Handles waiting until a specific response is provided by the server
+*/
+type await_response struct {
+	mu                sync.Mutex
+	requiredResponses map[string]string // halts write thread until the correlating response is provided by the server
+	haltedOn          string
+}
+
+/*
+Creates an active websocket connection
+*/
+func (tv_api *TV_API) OpenConnection() error {
+	// setting up the header
+	header := http.Header{}
+	header.Add("Origin", TV_ORIGIN_URL)
+
+	// creating the websocket
+	ws, _, err := websocket.DefaultDialer.Dial(TV_URL, header)
+	if err != nil {
+		return err
+	}
+
+	// initialise in values for the struct
+	tv_api.ws = ws
+
+	tv_api.ReadCh = make(chan map[string]interface{})
+	tv_api.writeCh = make(chan map[string]interface{})
+	tv_api.ErrorCh = make(chan error)
+	tv_api.internalErrorCh = make(chan error)
+
+	tv_api.symbolCounter = 0
+	tv_api.seriesCounter = 0
+	tv_api.seriesCreated = false // use sync.Once in clean up
+	tv_api.resolvedSymbols = make(map[string]string)
+	tv_api.seriesMap = make(map[string]string)
+
+	tv_api.csToken = "cs_" + createToken()
+	tv_api.qsToken = createToken()
+	tv_api.qs = "qs_" + tv_api.qsToken
+	tv_api.qssq = "qs_snapshoter_basic-symbol-quotes_" + tv_api.qsToken
+	tv_api.realtimeSymbols = make(map[string]bool)
+
+	// required responses for a given request being sent
+	// halts write requests from being sent until it is received
+	tv_api.halts = await_response{
+		requiredResponses: map[string]string{
+			"create_series":     "series_completed",
+			"modify_series":     "series_completed",
+			"request_more_data": "series_completed",
```

Comment:
`"resolve_symbol":    "symbol_resolved",`

fixes history not loading sometimes

Reply:
nice spot, I'll add it quickly